tsinfer/algorithm.py

```diff
@@ -1118,3 +1119,166 @@ def run_traceback(self, start, end, match):
+    # `active` is the list of ancestors that overlap with the current linesweep position.
+    active = [-1]
+    active.pop()
+    children = [[-1] for _ in range(n)]
+    for c in range(n):
+        children[c].pop()
+    incoming_edge_count = np.zeros(n, dtype=np.int32)
+    for i in range(len(event_times)):
+        index = event_index[i]
+        e_time = event_times[i]
+        if event_type[i] == 1:
+            for j in active:
+                if new_time[j] > e_time:
+                    incoming_edge_count[index] += 1
+                    children[j].append(index)
+                elif new_time[j] < e_time:
+                    incoming_edge_count[j] += 1
+                    children[index].append(j)
+            active.append(index)
+        else:
+            active.remove(index)
+
+    # Convert children to ragged array format so we can pass arrays to the
+    # next numba function, `find_groups`.
+    children_data = []
+    children_indices = [0]
+    for child_list in children:
+        children_data.extend(child_list)
+        children_indices.append(len(children_data))
+    children_data = np.array(children_data, dtype=np.int32)
+    children_indices = np.array(children_indices, dtype=np.int32)
+    return children_data, children_indices, incoming_edge_count
+
+
+@numba.njit
+def find_groups(children_data, children_indices, incoming_edge_count):
+    # We find groups of ancestors that can be matched in parallel by topologically
+    # sorting the dependency graph. We do this by deconstructing the graph, removing
+    # nodes with no incoming edges, and adding them to a group.
+    n = len(children_indices) - 1
+    group_id = np.full(n, -1, dtype=np.int32)
+    current_group = 0
+    while True:
+        # Find the nodes with no incoming edges
+        no_incoming = np.where(incoming_edge_count == 0)[0]
+        if len(no_incoming) == 0:
+            break
+        # Remove them from the graph
+        for i in no_incoming:
+            incoming_edge_count[i] = -1
+            incoming_edge_count[
+                children_data[children_indices[i] : children_indices[i + 1]]
+            ] -= 1
+        # Add them to the group
+        group_id[no_incoming] = current_group
+        current_group += 1
+    return group_id
+
+
+def group_ancestors_by_linesweep(start, end, time):
+    # For a given set of ancestors, we want to group them for matching in parallel.
+    # For each ancestor, any overlapping, older ancestors must be in an earlier group,
+    # and any overlapping, younger ancestors in a later group. Any overlapping same-age
+    # ancestors must be in the same group so they don't match to each other.
+    # We do this by first merging the overlapping same-age ancestors. Then build a
+    # dependency graph of the ancestors by linesweep. Then form groups by topological
+    # sort. Finally, we un-merge the same-age ancestors.
+
+    assert len(start) == len(end)
+    assert len(start) == len(time)
```

Comment:
Good to add a few ``logging.info`` comments in this function so that we can see what's going on and the relative timings of the steps

Reply:
Added logging for timing, and also info such as where the linesweep cutoff was.